File: Tests/executors/CXX/database/src/MultiUserExecutor.cpp
#include "MultiUserExecutor.h"

#include <Utility/WebFrameworkUtility.hpp>

MultiUserExecutor::MultiUserExecutor() :
    userId(framework::utility::uuid::generateUUID())
{
	
}

void MultiUserExecutor::doGet(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	framework::Table table = request.getTable("test_database", "multi_user");

	framework::SQLResult result = table.execute
	(
		"SELECT * FROM multi_user WHERE user_id = ?",
		framework::utility::database::makeSQLValues(userId)
	);
	std::vector<framework::JsonObject> data;

	for (const auto& value : result)
	{
		framework::JsonObject object;

		object["id"] = value.at("id").get<int64_t>();
		object["user_id"] = value.at("user_id").get<std::string>();
		object["data"] = value.at("data").get<std::string>();

		data.push_back(std::move(object));
	}

	response.setBody(framework::JsonBuilder().append("data", std::move(data)));
}

void MultiUserExecutor::doPost(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	request.getOrCreateDatabase("test_database").getOrCreateTable("multi_user", "CREATE TABLE IF NOT EXISTS multi_user (id INTEGER PRIMARY KEY AUTOINCREMENT, user_id TEXT NOT NULL, data TEXT NOT NULL)");
}

void MultiUserExecutor::doPut(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	framework::Table table = request.getTable("test_database", "multi_user");

	table.execute
	(
		"INSERT INTO multi_user (user_id, data) VALUES(?, ?)",
		framework::utility::database::makeSQLValues(userId, request.getJSON().get<std::string>("data"))
	);
}

DEFINE_EXECUTOR(MultiUserExecutor)

File: Tests/executors/CSharp/DatabaseExecutors/MultiUserExecutor.cs
using Framework;
using Framework.Utility;

public class MultiUserExecutor : HeavyOperationStatefulExecutor
{
	private readonly string userId;

	public MultiUserExecutor()
	{
		userId = Guid.NewGuid().ToString();
	}

	public override void DoGet(HttpRequest request, HttpResponse response)
	{
		Table table = request.GetTable("test_database", "multi_user");

		SqlResult result = table.ExecuteQuery
		(
			"SELECT * FROM multi_user WHERE user_id = ?",
			WebFrameworkUtility.MakeSqlValues(userId)
		);
		List<object> jsonData = [];

		foreach (var row in result)
		{
			jsonData.Add
			(
				new
				{
					id = row["id"].GetValue<int>(),
					user_id = row["user_id"].GetValue<string>(),
					data = row["data"].GetValue<string>()
				}
			);
		}

		response.SetBody
		(
			new
			{
				data = jsonData
			}
		);
	}

	public override void DoPost(HttpRequest request, HttpResponse response)
	{
		request.GetOrCreateDatabase("test_database").GetOrCreateTable("multi_user", "CREATE TABLE IF NOT EXISTS multi_user (id INTEGER PRIMARY KEY AUTOINCREMENT, user_id TEXT NOT NULL, data TEXT NOT NULL)");
	}

	public override void DoPut(HttpRequest request, HttpResponse response)
	{
		Table table = request.GetTable("test_database", "test_table");

		table.ExecuteQuery
		(
			"INSERT INTO multi_user (user_id, data) VALUES(?, ?)",
			WebFrameworkUtility.MakeSqlValues(userId, request.GetJson()["data"].GetString()!)
		);
	}
}
